drivers/sdhc/ifx_cat1_sdio.c

```diff
@@ -0,0 +1,346 @@
+
+#include <zephyr/irq.h>
+
+#define IFX_CAT1_SDIO_F_MIN (SDMMC_CLOCK_400KHZ)
+#define IFX_CAT1_SDIO_F_MAX (SD_CLOCK_50MHZ)
+
+struct ifx_cat1_sdio_config {
+	const struct pinctrl_dev_config *pincfg;
+	SDHC_Type *reg_addr;
+	uint8_t irq_priority;
+};
+
+struct ifx_cat1_sdio_data {
+	cyhal_sdio_t sdio_obj;
+	cyhal_resource_inst_t hw_resource;
+	cyhal_sdio_configurator_t cyhal_sdio_config;
+	enum sdhc_clock_speed clock_speed;
+	enum sdhc_bus_width bus_width;
+	struct k_mutex access_mutex;
+
+	void *sdio_cb_user_data;
+	sdhc_interrupt_cb_t sdio_cb;
+};
+
+static uint32_t sdio_rca;
+static const cy_stc_sd_host_init_config_t host_config = {false, CY_SD_HOST_DMA_ADMA2, false};
+static cy_en_sd_host_card_capacity_t sd_host_card_capacity = CY_SD_HOST_SDSC;
+static cy_en_sd_host_card_type_t sd_host_card_type = CY_SD_HOST_NOT_EMMC;
+static cy_stc_sd_host_sd_card_config_t sd_host_sd_card_config = {
+	.lowVoltageSignaling = false,
+	.busWidth = CY_SD_HOST_BUS_WIDTH_4_BIT,
+	.cardType = &sd_host_card_type,
+	.rca = &sdio_rca,
+	.cardCapacity = &sd_host_card_capacity,
+};
+
+/* List of available SDHC instances */
+static SDHC_Type *const IFX_CAT1_SDHC_BASE_ADDRESSES[CY_IP_MXSDHC_INSTANCES] = {
+#ifdef SDHC0
+	SDHC0,
+#endif /* ifdef SDHC0 */
+
+#ifdef SDHC1
+	SDHC1,
+#endif /* ifdef SDHC1 */
+};
+
+static int32_t _get_hw_block_num(SDHC_Type *reg_addr)
+{
+	uint32_t i;
+
+	for (i = 0u; i < CY_IP_MXSDHC_INSTANCES; i++) {
+		if (IFX_CAT1_SDHC_BASE_ADDRESSES[i] == reg_addr) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int ifx_cat1_sdio_reset(const struct device *dev)
+{
+	struct ifx_cat1_sdio_data *dev_data = dev->data;
+
+	cyhal_sdhc_software_reset((cyhal_sdhc_t *)&dev_data->sdio_obj);
+
+	return 0;
+}
+
+static int ifx_cat1_sdio_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	cy_rslt_t ret;
+	struct ifx_cat1_sdio_data *dev_data = dev->data;
+	cyhal_sdio_cfg_t config = {.frequencyhal_hz = ios->clock};
+
+	/* NOTE: Set bus width, set card power, set host signal voltage,
+	 * set I/O timing does not support in current version of driver
+	 */
+
+	/* Set host clock */
+	if ((dev_data->clock_speed != ios->clock) && (ios->clock != 0)) {
+
+		if ((ios->clock > IFX_CAT1_SDIO_F_MAX) || (ios->clock < IFX_CAT1_SDIO_F_MIN)) {
+			return -EINVAL;
+		}
+
+		ret = cyhal_sdio_configure(&dev_data->sdio_obj, &config);
+		if (ret != CY_RSLT_SUCCESS) {
+			return -ENOTSUP;
+		}
+
+		dev_data->clock_speed = ios->clock;
+	}
+
+	return 0;
+}
+
+static int ifx_cat1_sdio_card_busy(const struct device *dev)
+{
+	struct ifx_cat1_sdio_data *dev_data = dev->data;
+
+	return cyhal_sdio_is_busy(&dev_data->sdio_obj) ? 1 : 0;
+}
+
+static int ifx_cat1_sdio_request(const struct device *dev, struct sdhc_command *cmd,
+				 struct sdhc_data *data)
+{
+	struct ifx_cat1_sdio_data *dev_data = dev->data;
+	int ret;
+
+	/* Ensure we have exclusive access to SD card before sending request */
+	if (k_mutex_lock(&dev_data->access_mutex, K_MSEC(cmd->timeout_ms)) != 0) {
```

Comment:
You should not need to lock access here- the SD framework already handles this during init as well as during each card I/O operation with its own mutex

Reply:
updated